### PR TITLE
plan: prune cor cols in gby items.

### PR DESCRIPTION
**Pull request by hanfei1991**
cor cols is treated as constant in gby items.
@shenli @coocood @zimulala @tiancaiamao PTAL

### CODE REVIEW

File: plan/column_pruning.go [hunk 1]
```diff
@@ -100,8 +100,17 @@ func (p *Aggregation) PruneColumns(parentUsedCols []*expression.Column) {
 			selfUsedCols = append(selfUsedCols, expression.ExtractColumns(arg)...)
 		}
 	}
-	for _, expr := range p.GroupByItems {
-		selfUsedCols = append(selfUsedCols, expression.ExtractColumns(expr)...)
+	if len(p.GroupByItems) > 0 {
+		for i := len(p.GroupByItems) - 1; i >= 0; i-- {
+			cols := expression.ExtractColumns(p.GroupByItems[i])
+			if len(cols) == 0 {
```
Comment (shenli) [inline plan/column_pruning.go]: When will len(cols) be zero?

File: plan/column_pruning.go [hunk 2]
```diff
@@ -100,8 +100,17 @@ func (p *Aggregation) PruneColumns(parentUsedCols []*expression.Column) {
...
+				p.GroupByItems = append(p.GroupByItems[:i], p.GroupByItems[i+1:]...)
+			}
+			selfUsedCols = append(selfUsedCols, cols...)
```
Comment (shenli) [inline plan/column_pruning.go]: if len(cols) == 0, should we save this line?

File: plan/column_pruning.go (hunk 1 above)
Comment (hanfei1991) [inline plan/column_pruning.go]: Like group by '1' or group by a correlated column.

Comment (shenli) [inline plan/column_pruning.go]: Then we do no need to "selfUsedCols = append(selfUsedCols, cols...)"

**Comment by hanfei1991** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

File: plan/column_pruning.go [hunk 3]
```diff
@@ -100,8 +100,18 @@ func (p *Aggregation) PruneColumns(parentUsedCols []*expression.Column) {
 			selfUsedCols = append(selfUsedCols, expression.ExtractColumns(arg)...)
 		}
 	}
-	for _, expr := range p.GroupByItems {
-		selfUsedCols = append(selfUsedCols, expression.ExtractColumns(expr)...)
+	if len(p.GroupByItems) > 0 {
```
Comment (coocood) [inline plan/column_pruning.go]: This line can be removed.

File: plan/column_pruning.go [hunk 4]
```diff
@@ -100,8 +100,18 @@ func (p *Aggregation) PruneColumns(parentUsedCols []*expression.Column) {
...
+		for i := len(p.GroupByItems) - 1; i >= 0; i-- {
+			cols := expression.ExtractColumns(p.GroupByItems[i])
+			if len(cols) == 0 {
+				p.GroupByItems = append(p.GroupByItems[:i], p.GroupByItems[i+1:]...)
```
Comment (coocood) [inline plan/column_pruning.go]: How about just set `p.GroupByItems[i] = expression.One`? The line 103 can be removed. 
The code would be mush simpler.
And add comments about why replace it to one.

Comment (hanfei1991) [inline plan/column_pruning.go]: If we set i to `One`, it will cost more to encode.

Comment (coocood) [inline plan/column_pruning.go]: It's a very rare case, I think the saved cost doesn't worth the complexity.

Comment (hanfei1991) [inline plan/column_pruning.go]: I don't think it adds the complexity.

Comment (coocood) [inline plan/column_pruning.go]: Writers always underestimate the complexity.

Comment (coocood) [inline plan/column_pruning.go]: Writers always underestimate the complexity.

Comment (hanfei1991) [inline plan/column_pruning.go]: I can't agree

Comment (coocood) [inline plan/column_pruning.go]: Please the provide a benchmark result difference for the reduced group by items, and estimate the frequency this case will be run.
Then give a few example in our projects that have lower `(performance Improvement *  frequency) / complexity`.

Comment (hanfei1991) [inline plan/column_pruning.go]: I doesn't need a bench test. If it has massive group by items, it must be very inefficient.

Comment (hanfei1991) [inline plan/column_pruning.go]: If you think that is complex, I think everything is complex, every optimization is complex, every sql should be written as best form by user.

File: plan/column_pruning.go [hunk 5]
```diff
@@ -100,8 +100,18 @@ func (p *Aggregation) PruneColumns(parentUsedCols []*expression.Column) {
...
+			} else {
+				selfUsedCols = append(selfUsedCols, cols...)
+			}
+		}
+		if len(p.GroupByItems) == 0 {
+			p.GroupByItems = []expression.Expression{expression.One}
```
Comment (zimulala) [inline plan/column_pruning.go]: Add comments about why use `expression.One`.
Rest LGTM

**Comment by hanfei1991** [comment]
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Comment by zimulala** [comment]
@coocood

**Comment by shenli** [comment]
LGTM

**Pull request merged**